### PR TITLE
Update docker images

### PR DESCRIPTION
**Pull request by hanefi**
## Changes
- Add PG 12 support
- Introduce debian buster in build matrix

---------

## Deviating distro packaging steps
If we run `update_dockerfiles` we will start to see some extra steps in two distros as they deviated from others:
- Oracle Linux 7: added some extra steps because `From January 2019 the previous method of refreshing the Yum repositories has altered` (See https://oracle-base.com/articles/linux/download-the-latest-oracle-linux-repo-file#oracle-linux-7-updated)
- Debian Buster: Needed one extra `zlib1g-dev` package when releasing the latest `pg_auto_failover` version

-----------

## Distros without support (yet)
- el-ol/8
	- No `ol8_software_collections1` `ol8_developer` `ol8_developer_EPEL` repositories available yet. Needs a lot more changes to dockerfiles to install the packaging dependencies
- Ubuntu Trusty EOLd. No PGDG repositories for this distro

## Other problems:
- We used to install latest Citus packages to get header files. I do not understand the logic behind this, and I did not want to install PG11 in an image that will create PG12 packages.
- `make_pg_buildext_parallel.patch` no longer can be applied to the latest `pg_buildext` scripts. We need to update the patch to continue using it.
- We may need to check if the gpg key for the debian builds have expired. A comment suggests that it did : `pub   4096R/ACCC4CF8 2011-10-13 [expires: 2019-07-02]` in https://github.com/citusdata/packaging/pull/351/files/c3bae51258c5d89df5b4c6aaa15b8ae3ad9a5490#diff-0199464a84de2fb83ab569432ffa676fR25

### CODE REVIEW

File: dockerfiles/centos-6-pg12/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    && yum clean all
+
+# install jq to process JSON API responses
+RUN curl -sL https://github.com/stedolan/jq/releases/download/jq-1.5/jq-linux64 \
+         -o /usr/bin/jq \
+    && chmod +x /usr/bin/jq
+
+# DO NOT install latest Citus release to get header files
+RUN curl -s https://packagecloud.io/install/repositories/citusdata/community/script.rpm.sh | bash \
+    # && yum install -y citus80_11 \
```
Comment (velioglu) [inline dockerfiles/centos-6-pg12/Dockerfile]: Will we install it when we have 12 packages? If so, we may change the version. If it is unnecessary we may completely remove it.

File: dockerfiles/centos-7-pg12/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    && yum clean all
+
+# install jq to process JSON API responses
+RUN curl -sL https://github.com/stedolan/jq/releases/download/jq-1.5/jq-linux64 \
+         -o /usr/bin/jq \
+    && chmod +x /usr/bin/jq
+
+# DO NOT install latest Citus release to get header files
+RUN curl -s https://packagecloud.io/install/repositories/citusdata/community/script.rpm.sh | bash \
+    # && yum install -y citus83_11 \
```
Comment (velioglu) [inline dockerfiles/centos-7-pg12/Dockerfile]: Same with centos-6-pg-12 comment

File: dockerfiles/debian-buster-all/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    && chmod +x /usr/bin/jq
+
+# DO NOT install latest Citus release to get header files
+RUN curl -s https://packagecloud.io/install/repositories/citusdata/community/script.deb.sh | bash \
+#    && apt-get install -y postgresql-9.6-citus-8.0 postgresql-10-citus-8.0 postgresql-11-citus-8.0 \
+    && rm -rf /var/lib/apt/lists/*
+
+# FIXME: get the patches working again
+# patch pg_buildext to use multiple processors
+# COPY make_pg_buildext_parallel.patch /
```
Comment (velioglu) [inline dockerfiles/debian-buster-all/Dockerfile]: If we won't use this patch, we shouldn't add the patch

File: dockerfiles/centos-8-pg11/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,52 @@
+# vim:set ft=dockerfile:
```
Comment (velioglu) [inline dockerfiles/centos-8-pg11/Dockerfile:1]: We should either remove Centos 8 support for now or update the versions in this file.

File: dockerfiles/centos-8-pg12/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# vim:set ft=dockerfile:
+FROM centos:7
```
Comment (velioglu) [inline dockerfiles/centos-8-pg12/Dockerfile]: Same comment with centos-8-pg-11

File: dockerfiles/debian-buster-all/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+# Ubuntu includes "gnupg" (not "gnupg2", but still 2.x), but not dirmngr, and gnupg 2.x requires dirmngr
+# so, if we're not running gnupg 1.x, explicitly install dirmngr too
+	&& { \
+		gpg --version | grep -q '^gpg (GnuPG) 1\.' \
+		|| apt-get install -y --no-install-recommends dirmngr \
+	; } \
+	&& rm -rf /var/lib/apt/lists/*
+
+RUN set -ex; \
+# pub   4096R/ACCC4CF8 2011-10-13 [expires: 2019-07-02]
```
Comment (velioglu) [inline dockerfiles/debian-buster-all/Dockerfile]: It looks the key has expired.

File: dockerfiles/debian-buster-all/make_pg_buildext_parallel.patch [hunk 7]
```diff
@@ -0,0 +1,17 @@
+--- /usr/bin/pg_buildext	2018-08-21 07:58:46.000000000 +0000
```
Comment (velioglu) [inline dockerfiles/debian-buster-all/make_pg_buildext_parallel.patch]: If we can't apply it, we should remove the file.

File: dockerfiles/debian-jessie-all/Dockerfile [hunk 8]
```diff
@@ -65,14 +65,15 @@ RUN curl -sL https://github.com/stedolan/jq/releases/download/jq-1.5/jq-linux64
          -o /usr/bin/jq \
     && chmod +x /usr/bin/jq
 
-# install latest Citus release to get header files
+# DO NOT install latest Citus release to get header files
 RUN curl -s https://packagecloud.io/install/repositories/citusdata/community/script.deb.sh | bash \
-    && apt-get install -y postgresql-9.6-citus-8.0 postgresql-10-citus-8.0 postgresql-11-citus-8.0 \
+#    && apt-get install -y postgresql-9.6-citus-8.0 postgresql-10-citus-8.0 postgresql-11-citus-8.0 \
```
Comment (velioglu) [inline dockerfiles/debian-jessie-all/Dockerfile]: Same with centos-6-pg-12 comment

File: dockerfiles/debian-jessie-all/Dockerfile [hunk 9]
```diff
@@ -65,14 +65,15 @@ RUN curl -sL https://github.com/stedolan/jq/releases/download/jq-1.5/jq-linux64
...
     && rm -rf /var/lib/apt/lists/*
 
+# FIXME: get the patches working again
 # patch pg_buildext to use multiple processors
-COPY make_pg_buildext_parallel.patch /
-RUN patch `which pg_buildext` < /make_pg_buildext_parallel.patch
+# COPY make_pg_buildext_parallel.patch /
+# RUN patch `which pg_buildext` < /make_pg_buildext_parallel.patch
```
Comment (velioglu) [inline dockerfiles/debian-jessie-all/Dockerfile]: Why we can not apply the patch to the build_ext? If we are planning to solve it later, please don't forget to open an issue.

File: dockerfiles/oraclelinux-6-pg11/Dockerfile [hunk 10]
```diff
@@ -6,7 +6,8 @@ RUN ( yum install -y yum-plugin-ovl || yum install -y yum-plugin-ovl ) \
     && yum clean all
 
 # install build tools and PostgreSQL development files
```
Comment (velioglu) [inline dockerfiles/oraclelinux-6-pg11/Dockerfile:21]: Please add the reason of this change to the commit message.

File: dockerfiles/centos-8-pg11/Dockerfile (hunk 4 above)
Comment (hanefi) [reply dockerfiles/centos-8-pg11/Dockerfile:1]: removed the files

File: dockerfiles/centos-8-pg12/Dockerfile (hunk 5 above)
Comment (hanefi) [reply dockerfiles/centos-8-pg12/Dockerfile]: removed the files

**Comment by jasonmp85** [comment]
> We used to install latest Citus packages to get header files. I do not understand the logic behind this, and I did not want to install PG11 in an image that will create PG12 packages.

This was to be able to build the shard rebalancer, which was why I really wanted to make it part of Enterprise, somehow. Looks like that's done now!

File: dockerfiles/debian-buster-all/make_pg_buildext_parallel.patch (hunk 7 above)
Comment (jasonmp85) [reply dockerfiles/debian-buster-all/make_pg_buildext_parallel.patch]: I'll fix this; this is _mostly_ for having builds actually use all cores.

File: dockerfiles/centos-6-pg12/Dockerfile (hunk 1 above)
Comment (jasonmp85) [reply dockerfiles/centos-6-pg12/Dockerfile]: It's unnecessary so long as we don't expect to have to build the shard rebalancer as a standalone package any more.

File: os-list.csv [hunk 11]
```diff
@@ -1,5 +1,6 @@
 centos,7
 centos,6
+debian,buster
 debian,jessie
 debian,stretch
 fedora,28
```
Comment (jasonmp85) [inline os-list.csv]: Seems like removing Fedora also makes sense, since we don't release it anymore, IIRC…

Comment (jasonmp85) [inline os-list.csv]: Seems like removing Fedora would make sense, since we don't release it anymore, IIRC… Though I guess that could be a separate PR?

And what about the Trusty removal?

File: templates/Dockerfile-deb.tmpl [hunk 12]
```diff
@@ -57,22 +57,25 @@ RUN echo 'deb http://apt.postgresql.org/pub/repos/apt/ %%release%%-pgdg main' >
         libxslt-dev \
         lintian \
         postgresql-server-dev-all \
-        wget \
+        # Add 12 package explicitly as *-all does not yet depend on 12 yet
+        postgresql-server-dev-12 \
+                wget \
```
Comment (jasonmp85) [inline templates/Dockerfile-deb.tmpl]: Dunno what's going on with indentation here; shouldn't be a change at all.

File: templates/Dockerfile-deb.tmpl [hunk 13]
```diff
@@ -57,22 +57,25 @@ RUN echo 'deb http://apt.postgresql.org/pub/repos/apt/ %%release%%-pgdg main' >
         libxslt-dev \
         lintian \
         postgresql-server-dev-all \
-        wget \
+        # Add 12 package explicitly as *-all does not yet depend on 12 yet
```
Comment (jasonmp85) [inline templates/Dockerfile-deb.tmpl]: This line is hopefully temporary. Maybe rewrite as _As of Oct. 28, we still must add server-dev-12 as server-dev-all does not yet depend on it_. Sure, it'll be two lines but it'll explain the timeline better.

File: templates/Dockerfile-deb.tmpl [hunk 14]
```diff
@@ -57,22 +57,25 @@ RUN echo 'deb http://apt.postgresql.org/pub/repos/apt/ %%release%%-pgdg main' >
...
+        postgresql-server-dev-12 \
+                wget \
     && rm -rf /var/lib/apt/lists/*
 
 # install jq to process JSON API responses
 RUN curl -sL https://github.com/stedolan/jq/releases/download/jq-1.5/jq-linux64 \
          -o /usr/bin/jq \
     && chmod +x /usr/bin/jq
 
-# install latest Citus release to get header files
```
Comment (jasonmp85) [inline templates/Dockerfile-deb.tmpl:68]: This _entire_ `RUN` invocation should be removed. The header files were only needed in our build images for the old rebalancer package, and if we've retired that in favor of putting the functionality directly in Enterprise, there's no need for this command at all. Kill it (it takes a _long_ time, too, and brings in a lot of packages. Good riddance).

File: templates/Dockerfile-deb.tmpl [hunk 15]
```diff
@@ -57,22 +57,25 @@ RUN echo 'deb http://apt.postgresql.org/pub/repos/apt/ %%release%%-pgdg main' >
...
+# DO NOT install latest Citus release to get header files
 RUN curl -s https://packagecloud.io/install/repositories/citusdata/community/script.deb.sh | bash \
-    && apt-get install -y postgresql-9.6-citus-8.0 postgresql-10-citus-8.0 postgresql-11-citus-8.0 \
+#    && apt-get install -y postgresql-9.6-citus-8.0 postgresql-10-citus-8.0 postgresql-11-citus-8.0 \
     && rm -rf /var/lib/apt/lists/*
 
+# FIXME: get the patches working again
```
Comment (jasonmp85) [inline templates/Dockerfile-deb.tmpl]: I've pushed a commit to fix this patch, so you should just revert the change here.

File: update_images [hunk 16]
```diff
@@ -19,18 +19,18 @@ while read -r line; do
 
     if [[ "${os}" = 'debian' ]] || [[ "${os}" = 'ubuntu' ]]; then
         tag="${os}-${release}-all"
-        args+="build -t citusdata/packaging:${tag} -f ${dockerfiles_dir}/${tag}/Dockerfile .\n"
+        args+="build --rm=false -t citus/packaging:${tag} -f ${dockerfiles_dir}/${tag}/Dockerfile .\n"
```
Comment (jasonmp85) [inline update_images]: Can you explain the changes on this line? I don't understand why they're necessary.

File: dockerfiles/debian-buster-all/Dockerfile (hunk 6 above)
Comment (jasonmp85) [reply dockerfiles/debian-buster-all/Dockerfile]: OK, so I used a tool (available on Debian, at least) called `pgpdump` to print out details of the packets in this key…

```
root@56a50a6a9a82:~# pgpdump key.pub 
Old: Public Key Packet(tag 6)(525 bytes)
	Ver 4 - new
	Public key creation time - Thu Oct 13 20:19:14 UTC 2011
	Pub alg - RSA Encrypt or Sign(pub 1)
	RSA n(4096 bits) - ...
	RSA e(17 bits) - ...
Old: User ID Packet(tag 13)(28 bytes)
	User ID - PostgreSQL Debian Repository
Old: Signature Packet(tag 2)(573 bytes)
	Ver 4 - new
	Sig type - Positive certification of a User ID and Public Key packet(0x13).
	Pub alg - RSA Encrypt or Sign(pub 1)
	Hash alg - SHA256(hash 8)
	Hashed Sub: key flags(sub 27)(1 bytes)
		Flag - This key may be used to certify other keys
		Flag - This key may be used to sign data
	Hashed Sub: preferred symmetric algorithms(sub 11)(4 bytes)
		Sym alg - AES with 256-bit key(sym 9)
		Sym alg - AES with 192-bit key(sym 8)
		Sym alg - AES with 128-bit key(sym 7)
		Sym alg - CAST5(sym 3)
	Hashed Sub: preferred hash algorithms(sub 21)(4 bytes)
		Hash alg - SHA512(hash 10)
		Hash alg - SHA384(hash 9)
		Hash alg - SHA256(hash 8)
		Hash alg - SHA224(hash 11)
	Hashed Sub: preferred compression algorithms(sub 22)(4 bytes)
		Comp alg - ZLIB <RFC1950>(comp 2)
		Comp alg - BZip2(comp 3)
		Comp alg - ZIP <RFC1951>(comp 1)
		Comp alg - Uncompressed(comp 0)
	Hashed Sub: features(sub 30)(1 bytes)
		Flag - Modification detection (packets 18 and 19)
	Hashed Sub: key server preferences(sub 23)(1 bytes)
		Flag - No-modify
	Hashed Sub: signature creation time(sub 2)(4 bytes)
		Time - Sat Oct 13 08:49:48 UTC 2012
	Hashed Sub: key expiration time(sub 9)(4 bytes)
		Time - Sun Oct 13 08:49:37 UTC 2013
	Sub: issuer key ID(sub 16)(8 bytes)
		Key ID - 0x7FCC7D46ACCC4CF8
	Hash left 2 bytes - 6e 09 
	RSA m^d mod n(4095 bits) - ...
		-> PKCS-1
Old: Signature Packet(tag 2)(573 bytes)
	Ver 4 - new
	Sig type - Positive certification of a User ID and Public Key packet(0x13).
	Pub alg - RSA Encrypt or Sign(pub 1)
	Hash alg - SHA256(hash 8)
	Hashed Sub: key flags(sub 27)(1 bytes)
		Flag - This key may be used to certify other keys
		Flag - This key may be used to sign data
	Hashed Sub: preferred symmetric algorithms(sub 11)(4 bytes)
		Sym alg - AES with 256-bit key(sym 9)
		Sym alg - AES with 192-bit key(sym 8)
		Sym alg - AES with 128-bit key(sym 7)
		Sym alg - CAST5(sym 3)
	Hashed Sub: preferred hash algorithms(sub 21)(4 bytes)
		Hash alg - SHA512(hash 10)
		Hash alg - SHA384(hash 9)
		Hash alg - SHA256(hash 8)
		Hash alg - SHA224(hash 11)
	Hashed Sub: preferred compression algorithms(sub 22)(4 bytes)
		Comp alg - ZLIB <RFC1950>(comp 2)
		Comp alg - BZip2(comp 3)
		Comp alg - ZIP <RFC1951>(comp 1)
		Comp alg - Uncompressed(comp 0)
	Hashed Sub: features(sub 30)(1 bytes)
		Flag - Modification detection (packets 18 and 19)
	Hashed Sub: key server preferences(sub 23)(1 bytes)
		Flag - No-modify
	Hashed Sub: signature creation time(sub 2)(4 bytes)
		Time - Sun Feb 24 19:44:21 UTC 2013
	Hashed Sub: key expiration time(sub 9)(4 bytes)
		Time - Wed Feb 24 19:44:03 UTC 2016
	Sub: issuer key ID(sub 16)(8 bytes)
		Key ID - 0x7FCC7D46ACCC4CF8
	Hash left 2 bytes - 45 31 
	RSA m^d mod n(4095 bits) - ...
		-> PKCS-1
Old: Signature Packet(tag 2)(573 bytes)
	Ver 4 - new
	Sig type - Positive certification of a User ID and Public Key packet(0x13).
	Pub alg - RSA Encrypt or Sign(pub 1)
	Hash alg - SHA256(hash 8)
	Hashed Sub: key flags(sub 27)(1 bytes)
		Flag - This key may be used to certify other keys
		Flag - This key may be used to sign data
	Hashed Sub: preferred symmetric algorithms(sub 11)(4 bytes)
		Sym alg - AES with 256-bit key(sym 9)
		Sym alg - AES with 192-bit key(sym 8)
		Sym alg - AES with 128-bit key(sym 7)
		Sym alg - CAST5(sym 3)
	Hashed Sub: preferred hash algorithms(sub 21)(4 bytes)
		Hash alg - SHA512(hash 10)
		Hash alg - SHA384(hash 9)
		Hash alg - SHA256(hash 8)
		Hash alg - SHA224(hash 11)
	Hashed Sub: preferred compression algorithms(sub 22)(4 bytes)
		Comp alg - ZLIB <RFC1950>(comp 2)
		Comp alg - BZip2(comp 3)
		Comp alg - ZIP <RFC1951>(comp 1)
		Comp alg - Uncompressed(comp 0)
	Hashed Sub: features(sub 30)(1 bytes)
		Flag - Modification detection (packets 18 and 19)
	Hashed Sub: key server preferences(sub 23)(1 bytes)
		Flag - No-modify
	Hashed Sub: signature creation time(sub 2)(4 bytes)
		Time - Wed Jan 29 14:50:01 UTC 2014
	Hashed Sub: key expiration time(sub 9)(4 bytes)
		Time - Tue Jul  2 14:49:19 UTC 2019
	Sub: issuer key ID(sub 16)(8 bytes)
		Key ID - 0x7FCC7D46ACCC4CF8
	Hash left 2 bytes - ce 64 
	RSA m^d mod n(4095 bits) - ...
		-> PKCS-1
Old: Signature Packet(tag 2)(590 bytes)
	Ver 4 - new
	Sig type - Positive certification of a User ID and Public Key packet(0x13).
	Pub alg - RSA Encrypt or Sign(pub 1)
	Hash alg - SHA256(hash 8)
	Hashed Sub: key flags(sub 27)(1 bytes)
		Flag - This key may be used to certify other keys
		Flag - This key may be used to sign data
	Hashed Sub: preferred symmetric algorithms(sub 11)(4 bytes)
		Sym alg - AES with 256-bit key(sym 9)
		Sym alg - AES with 192-bit key(sym 8)
		Sym alg - AES with 128-bit key(sym 7)
		Sym alg - CAST5(sym 3)
	Hashed Sub: preferred hash algorithms(sub 21)(4 bytes)
		Hash alg - SHA512(hash 10)
		Hash alg - SHA384(hash 9)
		Hash alg - SHA256(hash 8)
		Hash alg - SHA224(hash 11)
	Hashed Sub: preferred compression algorithms(sub 22)(4 bytes)
		Comp alg - ZLIB <RFC1950>(comp 2)
		Comp alg - BZip2(comp 3)
		Comp alg - ZIP <RFC1951>(comp 1)
		Comp alg - Uncompressed(comp 0)
	Hashed Sub: features(sub 30)(1 bytes)
		Flag - Modification detection (packets 18 and 19)
	Hashed Sub: key server preferences(sub 23)(1 bytes)
		Flag - No-modify
	Hashed Sub: issuer fingerprint(sub 33)(21 bytes)
	 v4 -	Fingerprint - b9 7b 0a fc aa 1a 47 f0 44 f2 44 a0 7f cc 7d 46 ac cc 4c f8 
	Hashed Sub: signature creation time(sub 2)(4 bytes)
		Time - Wed Jan  4 14:35:43 UTC 2017
	Sub: issuer key ID(sub 16)(8 bytes)
		Key ID - 0x7FCC7D46ACCC4CF8
	Hash left 2 bytes - 04 0b 
	RSA m^d mod n(4096 bits) - ...
		-> PKCS-1
```

As you can see above, that 2019 expiration date was from a signature made in 2014. A subsequent 2017 signature has no expiration, so there shouldn't be problems here.

I'm going to ask upstream (the PostgreSQL Docker packagers) about this comment, since it's copy-pasted from them, but I feel safe saying just to erase the expiration part from our Dockerfile template, as it's no longer correct (and probably never was?)

File: dockerfiles/debian-buster-all/Dockerfile (hunk 3 above)
Comment (jasonmp85) [reply dockerfiles/debian-buster-all/Dockerfile]: I think I've fixed it.

File: dockerfiles/debian-jessie-all/Dockerfile (hunk 8 above)
Comment (jasonmp85) [reply dockerfiles/debian-jessie-all/Dockerfile]: @velioglu, these files are all generated from the product of a template file and a list of OSes, so it really only needs to be updated in one place (the templates) and regenerated…

File: dockerfiles/ubuntu-trusty-all/Dockerfile [hunk 17]
```diff
@@ -65,14 +65,15 @@ RUN curl -sL https://github.com/stedolan/jq/releases/download/jq-1.5/jq-linux64
          -o /usr/bin/jq \
```
Comment (jasonmp85) [inline dockerfiles/ubuntu-trusty-all/Dockerfile:74]: I thought you said Trusty was EoL'd? In that case, can't you just `git rm` this whole subdirectory?

File: templates/Dockerfile-rpm.tmpl [hunk 18]
```diff
@@ -6,7 +6,8 @@ RUN ( yum install -y yum-plugin-ovl || yum install -y yum-plugin-ovl ) \
     && yum clean all
 
 # install build tools and PostgreSQL development files
-RUN yum groupinstall -y 'Development Tools' \
+RUN yum install -y centos-release-scl-rh epel-release \
```
Comment (jasonmp85) [inline templates/Dockerfile-rpm.tmpl]: Can you explain this change? I don't understand why we need `epel` and `scl` when we've gotten by without them for so long. What changed in 12?

File: templates/Dockerfile-rpm.tmpl [hunk 19]
```diff
@@ -16,6 +17,8 @@ RUN yum groupinstall -y 'Development Tools' \
         libicu-devel \
         libxml2-devel \
         libxslt-devel \
+        llvm-toolset-7-clang \
```
Comment (jasonmp85) [inline templates/Dockerfile-rpm.tmpl]: Shouldn't the PostgreSQL development packages transitively depend on this stuff if supported on a given platform? I think these things became necessary with the addition of the JIT stuff, but I also was able to deal with it by just relying on the transitive dependencies for each platform where JIT was or wasn't supported, I think.

What error made this necessary?

**Review by jasonmp85 — CHANGES_REQUESTED** [review]
This mostly looks good; I don't have any worries about its functionality, just questions about why certain changes were made in cases that are unclear to me. With satisfactory answers I think this is pretty much good to go…

Change-wise, just revert the places you disabled that `patch` file's application: I've fixed it in a commit I pushed. You can also _entirely_ remove the installation of the Citus packagecloud repos, I think. They were necessary _only_ for building the rebalancer, and represented a significantly annoying bootstrapping problem in our packaging workflow I'm glad to be rid of.

Other than that, I mostly have questions about why you need to add extra `yum` repositories (such as `epel`), and why we need to now explicitly install the `llvm` packages on those platforms.

Additionally, the change from `citusdata` to `citus` and the use of the `rm=false` flag in your `docker build` invocation confuses me, so I'd like to know what made them necessary.

After that, we're probably at :shipit:!

File: dockerfiles/debian-buster-all/Dockerfile (hunk 6 above)
Comment (jasonmp85) [reply dockerfiles/debian-buster-all/Dockerfile]: Asked about this in docker-library/postgres#623.

Comment (jasonmp85) [reply dockerfiles/debian-buster-all/Dockerfile]: BTW a dev responded on that issue and confirmed the comment was stale: https://github.com/docker-library/postgres/issues/623#issuecomment-547673476

So we're good on this.

**Review by mtuncer — COMMENTED** [review]
After these changes I would like to add man page creation capability for pg_auto_failover packaging.  Current difficulty there is varying levels of sphinx support in different OSes.

File: templates/Dockerfile-deb.tmpl (hunk 15 above)
Comment (mtuncer) [reply templates/Dockerfile-deb.tmpl]: How important is this patch ? I remember removing this patch line when I was creating test packages locally.

File: os-list.csv (hunk 11 above)
Comment (hanefi) [reply os-list.csv]: I will remove EOLd distros in a seperate PR. In this PR I only want to introduce new distros. (Buster and CentOS 8)

File: templates/Dockerfile-deb.tmpl (hunk 13 above)
Comment (hanefi) [reply templates/Dockerfile-deb.tmpl]: Tested that the issue was temporary. Removing the extra lines and commments, as `postgresql-server-dev-all` is sufficient now

File: templates/Dockerfile-deb.tmpl (hunk 14 above)
Comment (hanefi) [reply templates/Dockerfile-deb.tmpl:68]: I am not so sure if `pg_auto_failover_enterprise` needs any citus packages. I am unsure, so I want to keep this here for now.

Updating the comments to include this reason in the file

File: templates/Dockerfile-deb.tmpl (hunk 15 above)
Comment (hanefi) [reply templates/Dockerfile-deb.tmpl]: It can speed things up. I will revert changes and update all the patches in relevant directories.

File: templates/Dockerfile-rpm.tmpl (hunk 18 above)
Comment (hanefi) [reply templates/Dockerfile-rpm.tmpl]: I saw that you introduced them in ff15bc394765d9a92cc054a344414679dfdb8fc1. The change was not made in a template, but in in the centos/oraclelinux pg11 docker files. To still be able to use a template I will try to add steps to install epel and scl in the template

File: templates/Dockerfile-rpm.tmpl (hunk 19 above)
Comment (hanefi) [reply templates/Dockerfile-rpm.tmpl]: We need llvm for JIT, and it was missing. The builds failed when `--with-llvm` was supplied

File: update_images (hunk 16 above)
Comment (hanefi) [reply update_images]: This change helped me recover faster during development. I recently reverted this change.
